velox/functions/prestosql/types/BigintEnumType.cpp

```diff
@@ -0,0 +1,130 @@
+  boost::replace_all(inputStr, "“", "\"");
+  boost::replace_all(inputStr, "”", "\"");
+
+  size_t colonPos = inputStr.find(':');
+  size_t parenthesisPos = inputStr.find('(', colonPos);
+  size_t braceOpen = inputStr.find('{', parenthesisPos);
+  size_t braceClose = inputStr.find('}', braceOpen);
+
+  if (colonPos == std::string::npos || parenthesisPos == std::string::npos ||
+      braceOpen == std::string::npos || braceClose == std::string::npos) {
+    throw std::invalid_argument("malformed enum type");
+  }
+
+  std::string type =
+      inputStr.substr(colonPos + 1, parenthesisPos - colonPos - 1);
+  boost::algorithm::trim(type);
+  if (type != "BigintEnum") {
+    throw std::invalid_argument("invalid type: " + type);
+  }
+
+  std::string name =
+      inputStr.substr(parenthesisPos + 1, braceOpen - parenthesisPos - 1);
+  boost::algorithm::trim(name);
+  enumName = boost::algorithm::to_upper_copy(name);
+
+  std::string mapBody =
+      inputStr.substr(braceOpen + 1, braceClose - braceOpen - 1);
+  boost::algorithm::trim(mapBody);
+
+  std::vector<std::string> entries;
+  boost::split(entries, mapBody, boost::is_any_of(","));
+
+  for (auto& entry : entries) {
+    boost::algorithm::trim(entry);
+    size_t quote1 = entry.find('"');
+    size_t quote2 = entry.find('"', quote1 + 1);
+    size_t colon = entry.find(':', quote2);
+
+    if (quote1 == std::string::npos || quote2 == std::string::npos ||
+        colon == std::string::npos) {
+      LOG(WARNING) << "invalid map entry: " << entry << "\n";
+      continue;
+    }
+
+    std::string key = entry.substr(quote1 + 1, quote2 - quote1 - 1);
+    std::string valStr = entry.substr(colon + 1);
+    boost::algorithm::trim(valStr);
+
+    try {
+      int64_t value = std::stoi(valStr);
+      enumMap[key] = value;
+    } catch (...) {
+      throw std::invalid_argument("invalid value " + valStr);
+    }
+  }
+  return {enumName, enumMap};
+}
+
+std::unordered_map<std::string, std::shared_ptr<const BigintEnumType>>&
+BigintEnumType::getInstances() {
+  static std::unordered_map<std::string, std::shared_ptr<const BigintEnumType>>
+      instances;
+  return instances;
+}
+
+// A new BigintEnumType is created for each enum type.
+// i.e. "test.enum.mood" is not the same type as "test.enum.languages"
+// Each instance of a BigintEnumType is stored in a static unordered map with
+// the enum names as keys, so if an instance for a given enum name already
+// exists, that instance will be returned. Otherwise, a new instance is created,
+// added to the map, then returned.
+std::shared_ptr<const BigintEnumType> BigintEnumType::create(
+    const std::string& enumName,
+    const std::unordered_map<std::string, int64_t>& enumMap) {
+  auto instance = BigintEnumType::get(enumName);
+  if (instance) {
+    return instance.value();
+  }
+
+  auto& instances = getInstances();
+  auto newInstance = std::make_shared<const BigintEnumType>(enumName, enumMap);
+  instances[enumName] = newInstance;
```

Comment:
is there size limit we could impose on this map, e.g. suppose there's a bug in the Presto coordinator and it starts randomly corrupting the enum names, since this map is static it would grow forever

It would be easier to debug if this had a hard limit, e.g. 10,000?, and failed with a clear error message, rather than having to wait for it to OOM and trying to debug the core dumps (or worse the workers just go into a slow decline as they limp along with limited memory available)